File: payjoin-ffi/java/src/test/java/org/payjoindevkit/TestHttp.java
package org.payjoindevkit;

import java.io.ByteArrayInputStream;
import java.net.InetSocketAddress;
import java.net.ProxySelector;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.security.KeyStore;
import java.security.cert.CertificateFactory;
import java.time.Duration;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;

/**
 * HTTP client for the v2 integration harness. A near-direct Java port of the Kotlin bindings'
 * {@code TestHttp.kt} (same class, same behavior) - that file is itself already close to plain
 * Java, so this only adjusts syntax, not approach.
 *
 * <p>The in-process directory serves HTTPS with a self-signed certificate from
 * {@code payjoin-test-utils} ({@code local_cert_key()}, SANs {@code localhost} and
 * {@code 0.0.0.0}). This client trusts that one certificate and nothing else, and sends every
 * request through the OHTTP relay as an HTTP proxy.
 */
final class TestHttp implements AutoCloseable {
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(30);
    private final HttpClient client;

    TestHttp(TestServices services) throws Exception {
        this.client = buildClient(services);
    }

    byte[] post(Request request) throws Exception {
        HttpRequest httpRequest = HttpRequest.newBuilder(URI.create(request.url()))
                .timeout(REQUEST_TIMEOUT)
                .header("Content-Type", request.contentType())
                .POST(HttpRequest.BodyPublishers.ofByteArray(request.body()))
                .build();
        HttpResponse<byte[]> response = client.send(httpRequest, HttpResponse.BodyHandlers.ofByteArray());
        int status = response.statusCode();
        if (status < 200 || status >= 300) {
            throw new IllegalStateException("HTTP " + status + " posting to " + request.url());
        }
        return response.body();
    }

    @Override
    public void close() {
        client.close();
    }

    private static HttpClient buildClient(TestServices services) throws Exception {
        Duration timeout = Duration.ofSeconds(30);
        URI relay = URI.create(services.ohttpRelayUrl());
        int port = relay.getPort() == -1 ? 80 : relay.getPort();
        return HttpClient.newBuilder()
                .connectTimeout(timeout)
                .sslContext(sslContextTrusting(services.cert()))
                .proxy(ProxySelector.of(new InetSocketAddress(relay.getHost(), port)))
                .build();
    }

    private static SSLContext sslContextTrusting(byte[] certDer) throws Exception {
        var cert = CertificateFactory.getInstance("X.509")
                .generateCertificate(new ByteArrayInputStream(certDer));
        KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
        keyStore.load(null, null);
        keyStore.setCertificateEntry("directory", cert);
        TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
        tmf.init(keyStore);
        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, tmf.getTrustManagers(), null);
        return sslContext;
    }
}

File: payjoin-ffi/kotlin/src/test/kotlin/org/payjoindevkit/TestHttp.kt
package org.payjoindevkit

import java.io.ByteArrayInputStream
import java.net.InetSocketAddress
import java.net.ProxySelector
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.security.KeyStore
import java.security.cert.CertificateFactory
import java.time.Duration
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManagerFactory

/**
 * HTTP client for the v2 integration harness.
 *
 * The in-process directory serves HTTPS with a self-signed certificate from
 * `payjoin-test-utils` (`local_cert_key()`, SANs `localhost` and `0.0.0.0`).
 * This client trusts that one certificate and nothing else, and sends every
 * request through the OHTTP relay as an HTTP proxy.
 */
class TestHttp(services: TestServices) : AutoCloseable {
    private val requestTimeout: Duration = Duration.ofSeconds(30)
    private val client: HttpClient = buildClient(services)

    fun post(request: Request): ByteArray {
        val httpRequest = HttpRequest.newBuilder(URI.create(request.url))
            .timeout(requestTimeout)
            .header("Content-Type", request.contentType)
            .POST(HttpRequest.BodyPublishers.ofByteArray(request.body))
            .build()
        val response = client.send(httpRequest, HttpResponse.BodyHandlers.ofByteArray())
        val status = response.statusCode()
        if (status < 200 || status >= 300) {
            throw IllegalStateException("HTTP $status posting to ${request.url}")
        }
        return response.body()
    }

    override fun close() {
        client.close()
    }

    companion object {
        private fun buildClient(services: TestServices): HttpClient {
            val timeout = Duration.ofSeconds(30)
            val relay = URI.create(services.ohttpRelayUrl())
            val port = if (relay.port == -1) 80 else relay.port
            return HttpClient.newBuilder()
                .connectTimeout(timeout)
                .sslContext(sslContextTrusting(services.cert()))
                .proxy(ProxySelector.of(InetSocketAddress(relay.host, port)))
                .build()
        }

        private fun sslContextTrusting(certDer: ByteArray): SSLContext {
            val cert = CertificateFactory.getInstance("X.509")
                .generateCertificate(ByteArrayInputStream(certDer))
            val keyStore = KeyStore.getInstance(KeyStore.getDefaultType())
            keyStore.load(null, null)
            keyStore.setCertificateEntry("directory", cert)
            val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
            tmf.init(keyStore)
            val sslContext = SSLContext.getInstance("TLS")
            sslContext.init(null, tmf.trustManagers, null)
            return sslContext
        }
    }
}

internal inline fun <T : Disposable, R> T.useDisposable(block: (T) -> R): R =
    try { block(this) } finally { destroy() }
